File: ex6.pl
% Defining the graph structure using facts.

edge(a, b).
edge(a, c).
edge(b, d).
edge(b, e).
edge(c, f).
edge(e, f).

% BFS predicate that initializes the search.

bfs(Start) :-
	bfs_visit([Start], []).

% bfs_visit/2 keeps track of the visited nodes and the queue.

bfs_visit([], _) :-
	!. % stop if there are no more nodes in the queue.

bfs_visit([Node | Queue], Visited) :-
	\+ member(Node, Visited), % Ensure Node is not visited
	write(Node), write(' '),  % print the current node
	findall(NextNode, (edge(Node, NextNode), \+ member(NextNode, Visited)), Neighbors),
	append(Queue, Neighbors, NewQueue),
	bfs_visit(NewQueue, [Node | Visited]).

% If a node is already visited, skip it.

bfs_visit([Node | Queue], Visited) :-
	member(Node, Visited),
	bfs_visit(Queue, Visited).

File: ex6.c
#include <stdio.h>
#include <stdlib.h>
#define MAX 10
int graph[MAX][MAX], visited[MAX];
char nodes[MAX];
void bfs(int start, int n) {
int queue[MAX], front = 0, rear = 0;
visited[start] = 1;
queue[rear++] = start;
while (front < rear) {
int current = queue[front++];
printf("%c ", nodes[current]);
for (int i = 0; i < n; i++) {
if (graph[current][i] == 1 && !visited[i]) {
visited[i] = 1;
queue[rear++] = i;
}
}
}
}
int main() {
int n, i, j;
printf("Enter number of nodes: ");
scanf("%d", &n);
printf("Enter labels of nodes (e.g. A B C ...):\n");
for (i = 0; i < n; i++) {
scanf(" %c", &nodes[i]);
}
printf("Enter adjacency matrix (%d x %d):\n", n, n);
for (i = 0; i < n; i++)
for (j = 0; j < n; j++)
scanf("%d", &graph[i][j]);
// Initialize visited array
for (i = 0; i < n; i++)
visited[i] = 0;
char startNode;
printf("Enter starting node: ");
scanf(" %c", &startNode);
int startIndex = -1;
for (i = 0; i < n; i++) {
if (nodes[i] == startNode) {
startIndex = i;
break; } }
if (startIndex == -1) {
printf("Invalid start node!\n");
return 1;
}
printf("BFS traversal starting from node %c:\n", startNode);
bfs(startIndex, n);
return 0;
}